Add jsonl support

kfigiela opened the pull request:
JSON support is great, but JSONL is very common in streaming cases (e.g. logs). 

```
$ cat test.jsonl
{"a": 1}
{"a": 2}
{"a": 42}

$ alasql "select * from jsonl('./test') where a > 1"
[
  {
    "a": 2
  },
  {
    "a": 42
  }
]
```

kfigiela committed:
Add jsonl support

mathiasrw commented on src/84from.js:
```diff
@@ -135,6 +135,25 @@ alasql.from.JSON = function (filename, opts, cb, idx, query) {
+	alasql.utils.loadFile(filename, !!cb, function (data) {
+		res = data.split(/\r?\n/);
+		// Remove last line if empty
+		if (res[res.length - 1] === '') {
```
It could be more than one line that is empty, or an empty line in the middle. I thin its also good to ignore lines with only space. I propose you add 

.map(line => line.trim()).filter(Boolean); to line 142 and remove line 144-146

mathiasrw commented on src/84from.js:
```diff
@@ -135,6 +135,25 @@ alasql.from.JSON = function (filename, opts, cb, idx, query) {
+			res.pop();
+		}
+		for (var i = 0, ilen = res.length; i < ilen; i++) {
+			res[i] = JSON.parse(res[i]);
```
Are we happy with a random error message, or should we provide more informative error message from a try-catch?

kfigiela replied:
Right, enriching with line number can be super useful in case of invalid input.

mathiasrw commented:
Love it. Good and solid functionality to be added. Thank you. 

Any chance you can add a test file that confirms its working (so we detect if we break that later)?

Is `.jsonl` the right name? What about `.ndjson` (new line delimited json) or I have even seen `.jsons` ?

mathiasrw commented:
@kfigiela ?

kfigiela commented:
@mathiasrw, thanks for reminder. I'll try to address those over the weekend and add tests.

As far as the extension/naming, according to [a blog post](https://www.hypertesto.me/en/blog/jsonl-and-ndjson/) these are "competing", but practically identical standards. We only parse, so as long as JSON.parse can handle it we should be good. I think, it makes sense to register them both, so either file extension works.

paulrutter commented:
https://duckdb.org/2023/03/03/json.html uses ndjson indeed, would be good to rename.

kfigiela committed:
autoExtFilename to allow longer file extensions

kfigiela committed:
Refactor JSONL parser, add NDJSON as an alias. Fix loadFile error handling for JSON as well.

kfigiela committed:
Implement basic tests for JSONL/NDJSON parsing

kfigiela commented:
On the other hand, Google BigQuery mentions JSONL everywhere. Nevertheless, we can have both aliases working and this is what I have implemented. Otherwise, I addressed the above comments and added some tests.

mathiasrw commented:
Just provide all the most used names (alasql.jsonl = alasql.ndjson = alsql.....)

mathiasrw commented:
Ahh - you already did :)

mathiasrw commented on src/84from.js:
```diff
@@ -131,10 +131,47 @@ alasql.from.JSON = function (filename, opts, cb, idx, query) {
+		err => {
+			throw new Error(err);
+		};
 	return res;
```
THanks

mathiasrw commented:
Doing a bit of performance testing. 

<img width="294" alt="image" src="https://github.com/AlaSQL/alasql/assets/1063454/7259f7a7-4f95-4f00-9349-a26f26c174c7">

its faster just to regex `\s*\r?\n\s*/` as it removes any empty lines in one swoop. Ill add this before merging.

mathiasrw committed:
Fix automated feedback about regex
